src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/Helpers.cs

```diff
@@ -203,46 +203,122 @@ public static object[] InstantiateAsArray(this IEnumerable<CustomAttributeData>
+        private static object? GetRawDefaultValue(IEnumerable<CustomAttributeData> customAttributes)
         {
-            // Legacy: If there are multiple default value attribute, the desktop picks one at random (and so do we...)
-            foreach (CustomAttributeData cad in customAttributes)
+            foreach (CustomAttributeData attributeData in customAttributes)
             {
-                Type attributeType = cad.AttributeType;
-                if (attributeType.IsSubclassOf(typeof(CustomConstantAttribute)))
+                Type attributeType = attributeData.AttributeType;
+                if (attributeType == typeof(DecimalConstantAttribute))
                 {
-                    if (raw)
-                    {
-                        foreach (CustomAttributeNamedArgument namedArgument in cad.NamedArguments)
-                        {
-                            if (namedArgument.MemberName.Equals("Value"))
-                            {
-                                defaultValue = namedArgument.TypedValue.Value;
-                                return true;
-                            }
-                        }
-                        defaultValue = null;
-                        return false;
-                    }
-                    else
+                    return GetRawDecimalConstant(attributeData);
+                }
+                else if (attributeType.IsSubclassOf(typeof(CustomConstantAttribute)))
+                {
+                    if (attributeType == typeof(DateTimeConstantAttribute))
                     {
-                        CustomConstantAttribute customConstantAttribute = (CustomConstantAttribute)(cad.Instantiate());
-                        defaultValue = customConstantAttribute.Value;
-                        return true;
+                        return GetRawDateTimeConstant(attributeData);
                     }
+                    return GetRawConstant(attributeData);
                 }
-                if (attributeType.Equals(typeof(DecimalConstantAttribute)))
+            }
+            return DBNull.Value;
+        }
+
+        private static decimal GetRawDecimalConstant(CustomAttributeData attr)
+        {
+            ParameterInfo[] parameters = attr.Constructor.GetParameters();
+            System.Collections.Generic.IList<CustomAttributeTypedArgument> args = attr.ConstructorArguments;
+
+            if (parameters[2].ParameterType == typeof(uint))
+            {
+                // DecimalConstantAttribute(byte scale, byte sign, uint hi, uint mid, uint low)
+                int low = (int)(uint)args[4].Value!;
+                int mid = (int)(uint)args[3].Value!;
+                int hi = (int)(uint)args[2].Value!;
+                byte sign = (byte)args[1].Value!;
+                byte scale = (byte)args[0].Value!;
+
+                return new decimal(low, mid, hi, sign != 0, scale);
```

Comment:
This could be simplified by not duplicating same code as bellow

Reply:
Agreed